src/Features/Core/Portable/AddImport/AbstractAddMissingImportsFeatureService.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.AddImport
+{
+    internal abstract class AbstractAddMissingImportsFeatureService : IAddMissingImportsFeatureService
+    {
+        private IDiagnosticService _diagnosticService;
+
+        public AbstractAddMissingImportsFeatureService(IDiagnosticService diagnosticService)
+        {
+            _diagnosticService = diagnosticService;
+        }
+
+        protected abstract AbstractAddImportCodeFixProvider GetAddImportCodeFixProvider();
+
+        public async Task<Project> AddMissingImportsAsync(Document document, CancellationToken cancellationToken)
+        {
+            var codeFixProvider = GetAddImportCodeFixProvider();
+
+            var project = document.Project;
+            var workspace = project.Solution.Workspace;
+            var diagnosticData = _diagnosticService.GetDiagnostics(workspace, project.Id, document.Id, id: null,
```

Comment:
you should use IDiagnosticAnalyzerService rather than IDiagnosticService. IDiagnosticService is approximate service that return last known data rather than exact data from the snapshot.

you should use IDiagnosticAnalyzerService that accepts solution/project/document and return exact diagnostics on the given snapshot.